blockchain/stake/internal/tickettreap/common.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2015-2016 The btcsuite developers
+// Copyright (c) 2016 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package tickettreap
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+const (
+	// staticDepth is the size of the static array to use for keeping track
+	// of the parent stack during treap iteration.  Since a treap has a very
+	// high probability that the tree height is logarithmic, it is
+	// exceedingly unlikely that the parent stack will ever exceed this size
+	// even for extremely large numbers of items.
+	staticDepth = 128
+
+	// nodeFieldsSize is the size the fields of each node takes excluding
+	// the contents of the key and value.  It assumes 64-bit pointers so
+	// technically it is smaller on 32-bit platforms, but overestimating the
+	// size in that case is acceptable since it avoids the need to import
+	// unsafe.  It consists of 8 bytes for each of the value, priority,
+	// left, and right fields (8*4).
+	nodeFieldsSize = 32
+
+	// nodeValueSize is the size of the fixed-size fields of a Value.
+	nodeValueSize = 4
+)
+
+var (
+	// rng is a new random number generator that is local to the package.
+	rng = rand.New(rand.NewSource(time.Now().UnixNano()))
+)
+
+// Key defines the key used to add an associated value to the treap.
+type Key chainhash.Hash
```

Comment:
Agreed about performance and I'd rather avoid that hit.

@cjepson can chime in here, but if there is no intention of ever keying off anything except the hash (which does seem pretty reasonable to me), this could be changed to just take `chainhash.Hash` everywhere instead of Key.

That said, there is no performance hit to type casting a `chainhash.Hash` to a `tickettreap.Key` either and the type is easier to update if it ever needs to be.